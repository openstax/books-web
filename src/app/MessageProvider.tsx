import { shouldPolyfill } from '@formatjs/intl-pluralrules/should-polyfill';
import React from 'react';
import { IntlProvider } from 'react-intl';
import enMessages from './messages/en';

// https://formatjs.io/docs/polyfills/intl-pluralrules/#dynamic-import--capability-detection
async function polyfill(locale: 'en') {
  if (shouldPolyfill()) {
    await import('@formatjs/intl-pluralrules/polyfill');
  }

  // boolean added by the polyfill
  if ((Intl.PluralRules as (typeof Intl.PluralRules & {polyfilled?: boolean})).polyfilled) {
    switch (locale) {
      case 'en':
        await import('@formatjs/intl-pluralrules/locale-data/en');
        break;
      default:
        throw new Error('unsupported locale in messageProvider');
    }
  }
}

polyfill('en');

// tslint:disable-next-line:variable-name
const MessageProvider: React.SFC<{onError?: () => void}> = (props) =>
  <IntlProvider onError={props.onError} locale='en' messages={enMessages}>{props.children}</IntlProvider>;

export default MessageProvider;
