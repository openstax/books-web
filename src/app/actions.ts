import { createStandardAction } from 'typesafe-actions';
import { Experiments } from '../helpers/pollUpdates';

export const receiveFeatureFlags = createStandardAction('app/receiveFeatureFlags')<string[]>();
export const receivePageFocus = createStandardAction('app/receivePageFocus')<boolean>();
export const receiveExperiments = createStandardAction('app/receiveExperiments')<Experiments>();
