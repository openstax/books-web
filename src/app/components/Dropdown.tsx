import { HTMLElement } from '@openstax/types/lib.dom';
import flow from 'lodash/fp/flow';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled, { css, keyframes } from 'styled-components/macro';
import { useOnClickOutside } from '../content/highlights/components/utils/onClickOutside';
import theme from '../theme';
import { preventDefault, useOnEsc } from '../utils';
import { textStyle } from './Typography/base';

interface ToggleProps<T extends React.ComponentType = React.ComponentType> {
  className?: string;
  component: T extends React.ComponentType
    ? React.ReactComponentElement<T>:
    never;
}
// tslint:disable-next-line:variable-name
export const DropdownToggle = styled(({component, ...props}: ToggleProps) => React.cloneElement(component, props))`
  cursor: pointer;
`;

const fadeIn = keyframes`
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
`;

const fadeInAnimation = css`
  animation: ${100}ms ${fadeIn} ease-out;
`;

const visuallyShown = css`
  height: unset;
  width: unset;
  clip: unset;
  overflow: visible;
`;

const visuallyHidden = css`
  height: 0;
  width: 0;
  overflow: hidden;
  clip: rect(1px, 1px, 1px, 1px);
`;

type Props = React.PropsWithChildren<{
  toggle: React.ReactNode;
  className?: string;
}>;

// tslint:disable-next-line:variable-name
const TabHiddenDropDown = styled(({toggle, children, className}: Props) => {
  const [open, setOpen] = React.useState<boolean>(false);
  const container = React.useRef<HTMLElement>(null);

  useOnClickOutside(container, open, () => setOpen(false));
  useOnEsc(container, open, () => setOpen(false));

  return <div className={className} ref={container}>
    <DropdownToggle component={toggle} onClick={() => setOpen(!open)} />
    {open && children}
  </div>;
})`
  ${css`
    & > *:not(${DropdownToggle}) {
      ${fadeInAnimation}
      position: absolute;
      box-shadow: 0 0.5rem 0.5rem 0 rgba(0, 0, 0, 0.1);
      border: 1px solid ${theme.color.neutral.formBorder};
      top: calc(100% + 0.4rem);
      left: 0;
    }
  `}
`;

// tslint:disable-next-line:variable-name
const DropdownFocusWrapper = styled.div`
  overflow: visible;
`;

// tslint:disable-next-line:variable-name
const TabTransparentDropdown = styled(({toggle, children, className}: Props) => <div className={className}>
  <DropdownFocusWrapper>
    <DropdownToggle tabIndex={0} component={toggle} />
    {children}
  </DropdownFocusWrapper>
  <DropdownToggle tabIndex={0} component={toggle} />
</div>)`
  ${/* i don't know why stylelint was complaining about this but it was, css wrapper suppresses */ css`
    ${DropdownFocusWrapper} + ${DropdownToggle} {
      ${visuallyHidden}
    }
    ${DropdownFocusWrapper}.focus-within + ${DropdownToggle} {
      ${visuallyShown}
    }
    ${DropdownFocusWrapper}:focus-within + ${DropdownToggle} {
      ${visuallyShown}
    }

    ${DropdownFocusWrapper} > ${DropdownToggle} {
      ${visuallyShown}
    }
    ${DropdownFocusWrapper}.focus-within > ${DropdownToggle} {
      ${visuallyHidden}
    }
    ${DropdownFocusWrapper}:focus-within > ${DropdownToggle} {
      ${visuallyHidden}
    }

    ${DropdownFocusWrapper} > *:not(${DropdownToggle}) {
      ${fadeInAnimation}
      position: absolute;
      box-shadow: 0 0.5rem 0.5rem 0 rgba(0, 0, 0, 0.1);
      border: 1px solid ${theme.color.neutral.formBorder};
      top: calc(100% + 0.4rem);
      left: 0;
    }

    ${DropdownFocusWrapper} > *:not(${DropdownToggle}) {
      ${visuallyHidden}
    }

    ${DropdownFocusWrapper}.focus-within > *:not(${DropdownToggle}) {
      ${visuallyShown}
    }

    ${DropdownFocusWrapper}:focus-within > *:not(${DropdownToggle}) {
      ${visuallyShown}
    }
  `}
`;

// tslint:disable-next-line:variable-name
export const DropdownList = styled.ol`
  margin: 0;
  padding: 0.6rem 0;
  background: ${theme.color.neutral.formBackground};

  li {
    display: inline-block;
  }

  li button,
  li a {
    text-decoration: none;
    display: flex;
    align-items: center;
    text-align: left;
    cursor: pointer;
    outline: none;
    border: none;
    padding: 0 0.8rem;
    margin: 0;
    height: 3.2rem;
    background: none;
    min-width: 7.2rem;
    ${textStyle}
    font-size: 1.4rem;
    line-height: 2rem;

    &:hover,
    &:focus {
      background: ${theme.color.neutral.formBorder};
    }
  }
`;

// tslint:disable-next-line:variable-name
export const DropdownItem = ({message, href, onClick}: {message: string, href?: string, onClick?: () => void}) => <li>
  <FormattedMessage id={message}>
    {(msg: Element | string) => href
      ? <a href={href} onClick={onClick}>{msg}</a>
      /*
        this should be a button but Safari and firefox don't support focusing buttons
        which breaks the tab transparent dropdown
        https://bugs.webkit.org/show_bug.cgi?id=22261
        https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
      */
      // eslint-disable-next-line jsx-a11y/anchor-is-valid
      : <a tabIndex={0} href='' onClick={onClick ? flow(preventDefault, onClick) : preventDefault}>{msg}</a>
    }
  </FormattedMessage>
</li>;

// tslint:disable-next-line:variable-name
const Dropdown = ({transparentTab, ...props}: {transparentTab?: boolean} & Props) => transparentTab !== false
  ? <TabTransparentDropdown {...props} />
  : <TabHiddenDropDown {...props} />;

export default styled(Dropdown)`
  overflow: visible;
  position: relative;
`;
