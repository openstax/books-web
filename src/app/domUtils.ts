import { HTMLElement, Node, TouchEvent } from '@openstax/types/lib.dom';
import { isHtmlElement } from './guards';
import { assertWindow } from './utils';

export const SCROLL_UP: 'scroll_up' = 'scroll_up';
export const SCROLL_DOWN: 'scroll_down' = 'scroll_down';

export const getTouchDirection = (last: TouchEvent, next: TouchEvent) =>
  next.touches[0].clientY > last.touches[0].clientY
    ? SCROLL_UP
    : next.touches[0].clientY < last.touches[0].clientY
      ? SCROLL_DOWN
      : null
;

export const getTouchScrollElement = (last: TouchEvent, next: TouchEvent) => {
  let element = isHtmlElement(next.target) ? next.target : null;
  const direction = getTouchDirection(last, next);

  if (!direction) {
    return null;
  }

  const condition = direction === SCROLL_UP
    ? elementCanScrollUp
    : elementCanScrollDown
  ;

  while (element) {
    element = findFirstScrollableParent(element);

    if (element && condition(element)) {
      return element;
    } else if (element) {
      element = element.parentElement;
    }
  }

  return assertWindow();
};

export const elementCanScrollUp = (element: HTMLElement) =>
  element.scrollTop > 0;

export const elementCanScrollDown = (element: HTMLElement) =>
  element.offsetHeight + element.scrollTop < element.scrollHeight;

export const findFirstScrollableParent = (element: HTMLElement | null): HTMLElement | null => {

  const computedStyle = (el: HTMLElement) => assertWindow().getComputedStyle(el);
  const scrollyOverflows: Array<string | null> = ['auto', 'scroll'];

  if (
    !element ||
    (
      element.scrollHeight > element.offsetHeight &&
      scrollyOverflows.includes(computedStyle(element).overflow)
    )
  ) {
    return element;
  }

  return findFirstScrollableParent(element.parentElement);
};

export const findElementSelfOrParent = (node: Node) => {
  if (isHtmlElement(node)) {
    return node;
  } else if (node && node.parentElement) {
    return node.parentElement;
  }
};
