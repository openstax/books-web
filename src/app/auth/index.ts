import * as actions from './actions';
import * as hooks from './hooks';
import * as init from './init';
import reducer from './reducer';
import * as types from './types';

export {
  init,
  types,
  reducer,
  actions,
  hooks,
};
