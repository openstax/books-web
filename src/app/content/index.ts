import * as actions from './actions';
import * as hooks from './hooks';
import reducer from './reducer';
import * as routes from './routes';
import * as types from './types';

export {
  hooks,
  types,
  routes,
  reducer,
  actions,
};
