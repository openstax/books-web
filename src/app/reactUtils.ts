import { FocusEvent, HTMLElement, HTMLElementEventMap, KeyboardEvent } from '@openstax/types/lib.dom';
import React from 'react';
import { addSafeEventListener } from './domUtils';
import { isElement, isWindow } from './guards';
import { assertDefined } from './utils';

export const useDrawFocus = <E extends HTMLElement = HTMLElement>() => {
  const ref = React.useRef<E | null>(null);

  React.useEffect(() => {
    if (ref && ref.current) {
      ref.current.focus();
    }
  }, [ref]);

  return ref;
};

export const onFocusLostHandler = (ref: React.RefObject<HTMLElement>, isEnabled: boolean, cb: () => void) => () => {
  const el = ref && ref.current;
  if (!el) { return; }

  const handler = (event: FocusEvent) => {
    const relatedTarget = event.relatedTarget;

    if (!isElement(relatedTarget) || !ref.current!.contains(relatedTarget)) {
      cb();
    }
  };

  if (isEnabled) {
    return addSafeEventListener(el, 'focusout', handler);
  }
};

export const useFocusLost = (ref: React.RefObject<HTMLElement>, isEnabled: boolean, cb: () => void) => {
  React.useEffect(onFocusLostHandler(ref, isEnabled, cb), [ref, isEnabled]);
};

export const onDOMEventHandler = (
  element: React.RefObject<HTMLElement> | Window,
  isEnabled: boolean,
  event: keyof HTMLElementEventMap,
  cb: () => void
) => () => {
  const target = isWindow(element) ? element : element.current;

  if (!target) { return; }

  if (isEnabled) {
    target.addEventListener(event, cb);
  }

  return () => target.removeEventListener(event, cb);
};

export const useOnDOMEvent = (
  element: React.RefObject<HTMLElement> | Window ,
  isEnabled: boolean,
  event: keyof HTMLElementEventMap,
  cb: () => void,
  deps: React.DependencyList = []
) => {
  React.useEffect(onDOMEventHandler(element, isEnabled, event, cb), [element, isEnabled, event, cb, ...deps]);
};

export const useTimeout = (delay: number, callback: () => void) => {
  const savedCallback = React.useRef<typeof callback>();
  const timeout = React.useRef<number>();

  const timeoutHandler = () => savedCallback.current && savedCallback.current();
  const reset = () => {
    if (timeout.current) {
      clearTimeout(timeout.current);
    }

    timeout.current = setTimeout(timeoutHandler, delay);
  };

  React.useEffect(() => {
    savedCallback.current = callback;
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [callback]);

  React.useEffect(() => {
      reset();
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [delay]);

  React.useEffect(() => () => clearTimeout(assertDefined(timeout.current, 'timeout ID can\'t be undefined')), []);

  return reset;
};

/**
 * This function will return array where first item is a function which will set
 * event listener for given element and second item is a function which will remove
 * this listener.
 *
 * This function can be used in React class components.
 */
export const onEsc = (
  element: HTMLElement, cb: () => void
): [() => void, () => void] => {
  const handler = (e: KeyboardEvent) => {
    if (e.key === 'Escape') {
      e.stopPropagation();
      cb();
    }
  };

  return [
    () => element.addEventListener('keydown', handler),
    () => element.removeEventListener('keydown', handler),
  ];
};

export const onEscHandler = (element: React.RefObject<HTMLElement>, isEnabled: boolean, cb: () => void) => () => {
  const el = element && element.current;
  if (!el) { return; }

  const [addEvListener, removeEvListener] = onEsc(el, cb);
  if (isEnabled) {
    addEvListener();
  }

  return removeEvListener;
};

export const useOnEsc = (element: React.RefObject<HTMLElement>, isEnabled: boolean, cb: () => void) => {
  React.useEffect(onEscHandler(element, isEnabled, cb), [element, isEnabled]);
};
