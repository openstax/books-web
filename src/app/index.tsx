import { createBrowserHistory, createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { combineReducers } from 'redux';
import createStore from '../helpers/createStore';
import { PromiseCollector } from '../helpers/PromiseCollector';
import * as content from './content';
import * as errors from './errors';
import * as navigation from './navigation';
import { AnyAction, AppState, Middleware } from './types';

export const actions = {
  content: content.actions,
  errors: errors.actions,
  navigation: navigation.actions,
};

export const routes = [
  ...Object.values(content.routes),
  ...Object.values(errors.routes),
];

const hooks = [
  ...Object.values(content.hooks),
];

interface Options {
  initialState?: AppState;
  initialEntries?: any;
}

export default (options: Options = {}) => {
  const history = typeof window !== 'undefined' && window.history
    ? createBrowserHistory()
    : createMemoryHistory({initialEntries: options.initialEntries});

  const reducer = combineReducers<AppState, AnyAction>({
    content: content.reducer,
    errors: errors.reducer,
    navigation: navigation.createReducer(history.location),
  });

  const hookWatcher = new PromiseCollector();

  const middleware: Middleware[] = [
    navigation.createMiddleware(routes, history),
    ...hooks.map((hook) => hook(hookWatcher)),
  ];

  const store = createStore({
    middleware,
    reducer,
  });

  const container = () => <Provider store={store}>
    <navigation.components.NavigationProvider routes={routes} />
  </Provider>;

  navigation.utils.init(routes, history.location, store.dispatch);

  return {
    container,
    history,
    hooks: hookWatcher,
    store,
  };
};
