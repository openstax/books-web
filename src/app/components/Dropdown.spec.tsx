import React from 'react';
import renderer from 'react-test-renderer';
import * as reactUtils from '../../app/reactUtils';
import MessageProvider from '../MessageProvider';
import Dropdown, { DropdownItem, DropdownList } from './Dropdown';

describe('Dropdown', () => {
  it('matches snapshot', () => {
    const component = renderer.create(<MessageProvider>
      <Dropdown toggle={<button>show more</button>}>
        <DropdownList>
          <DropdownItem onClick={() => null} message='i18n:highlighting:dropdown:delete' />
          <DropdownItem onClick={() => null} href='/wooo' message='i18n:highlighting:dropdown:edit' />
        </DropdownList>
      </Dropdown>
    </MessageProvider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot for tab hidden (closed)', () => {
    const component = renderer.create(<MessageProvider>
      <Dropdown transparentTab={false} toggle={<button>show more</button>}>
        <DropdownList>
          <DropdownItem onClick={() => null} message='i18n:highlighting:dropdown:delete' />
          <DropdownItem onClick={() => null} href='/wooo' message='i18n:highlighting:dropdown:edit' />
        </DropdownList>
      </Dropdown>
    </MessageProvider>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot for tab hidden (open)', () => {
    const component = renderer.create(<MessageProvider>
      <Dropdown transparentTab={false} toggle={<button>show more</button>}>
        <DropdownList>
          <DropdownItem onClick={() => null} message='i18n:highlighting:dropdown:delete' />
          <DropdownItem onClick={() => null} href='/wooo' message='i18n:highlighting:dropdown:edit' />
        </DropdownList>
      </Dropdown>
    </MessageProvider>);

    renderer.act(() => {
      component.root.findByType('button').props.onClick();
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('tab hidden closes on focus lost', () => {
    const useFocusLost = jest.spyOn(reactUtils, 'useFocusLost');

    const component = renderer.create(<MessageProvider>
      <Dropdown transparentTab={false} toggle={<button>show more</button>}>
        <DropdownList>
          <DropdownItem onClick={() => null} message='i18n:highlighting:dropdown:delete' />
          <DropdownItem onClick={() => null} href='/wooo' message='i18n:highlighting:dropdown:edit' />
        </DropdownList>
      </Dropdown>
    </MessageProvider>);

    renderer.act(() => {
      component.root.findByType('button').props.onClick();
    });

    expect(() => component.root.findByType(DropdownList)).not.toThrow();

    renderer.act(() => {
      useFocusLost.mock.calls[0][2]();
    });

    expect(() => component.root.findByType(DropdownList)).toThrow();
  });

  it('tab hidden closes on Esc', () => {
    const useOnEscSpy = jest.spyOn(reactUtils, 'useOnEsc');

    const component = renderer.create(<MessageProvider>
      <Dropdown transparentTab={false} toggle={<button>show more</button>}>
        <DropdownList>
          <DropdownItem onClick={() => null} message='i18n:highlighting:dropdown:delete' />
          <DropdownItem onClick={() => null} href='/wooo' message='i18n:highlighting:dropdown:edit' />
        </DropdownList>
      </Dropdown>
    </MessageProvider>);

    renderer.act(() => {
      component.root.findByType('button').props.onClick();
    });

    expect(() => component.root.findByType(DropdownList)).not.toThrow();

    renderer.act(() => {
      useOnEscSpy.mock.calls[0][2]();
    });

    expect(() => component.root.findByType(DropdownList)).toThrow();

    useOnEscSpy.mockClear();
  });

  it('tab hidden focus after Esc', () => {
    const useOnEscSpy = jest.spyOn(reactUtils, 'useOnEsc');

    const focus = jest.fn();
    const addEventListener = jest.fn();
    const removeEventListener = jest.fn();
    const createNodeMock = () => ({focus, addEventListener, removeEventListener});

    const component = renderer.create(<MessageProvider>
      <Dropdown transparentTab={false} toggle={<button>show more</button>}>
        <DropdownList>
          <DropdownItem onClick={() => null} message='i18n:highlighting:dropdown:delete' />
          <DropdownItem onClick={() => null} href='/wooo' message='i18n:highlighting:dropdown:edit' />
        </DropdownList>
      </Dropdown>
    </MessageProvider>, {createNodeMock});

    renderer.act(() => {
      component.root.findByType('button').props.onClick();
    });

    expect(() => component.root.findByType(DropdownList)).not.toThrow();

    renderer.act(() => {
      useOnEscSpy.mock.calls[0][2]();
    });

    expect(() => component.root.findByType(DropdownList)).toThrow();
    expect(focus).toHaveBeenCalled();

    useOnEscSpy.mockClear();
  });
});
