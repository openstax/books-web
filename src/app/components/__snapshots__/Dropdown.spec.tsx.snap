// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown matches snapshot 1`] = `
.c5 {
  cursor: pointer;
}

.c3 {
  overflow: visible;
}

.c1 .c2 + .c4 {
  height: 0;
  width: 0;
  overflow: hidden;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c1 .c2.focus-within + .c4 {
  height: unset;
  width: unset;
  -webkit-clip: unset;
  clip: unset;
  overflow: visible;
}

.c1 .c2:focus-within + .c4 {
  height: unset;
  width: unset;
  -webkit-clip: unset;
  clip: unset;
  overflow: visible;
}

.c1 .c2 > .c4 {
  height: unset;
  width: unset;
  -webkit-clip: unset;
  clip: unset;
  overflow: visible;
}

.c1 .c2.focus-within > .c4 {
  height: 0;
  width: 0;
  overflow: hidden;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c1 .c2:focus-within > .c4 {
  height: 0;
  width: 0;
  overflow: hidden;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c1 .c2 > *:not(.c4) {
  -webkit-animation: 100ms bcCCNc ease-out;
  animation: 100ms bcCCNc ease-out;
  position: absolute;
  box-shadow: 0 0.5rem 0.5rem 0 rgba(0,0,0,0.1);
  border: 1px solid #d5d5d5;
  top: calc(100% + 0.4rem);
  left: 0;
}

.c1 .c2 > *:not(.c4) {
  height: 0;
  width: 0;
  overflow: hidden;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c1 .c2.focus-within > *:not(.c4) {
  height: unset;
  width: unset;
  -webkit-clip: unset;
  clip: unset;
  overflow: visible;
}

.c1 .c2:focus-within > *:not(.c4) {
  height: unset;
  width: unset;
  -webkit-clip: unset;
  clip: unset;
  overflow: visible;
}

.c6 {
  margin: 0;
  padding: 0.6rem 0;
  background: #f5f5f5;
}

.c6 li button,
.c6 li a {
  text-align: left;
  cursor: pointer;
  outline: none;
  border: none;
  padding: 0 0.8rem;
  margin: 0;
  height: 3.2rem;
  background: none;
  min-width: 7.2rem;
  font-family: Neue Helvetica W01;
  color: #424242;
  font-size: 1.4rem;
  line-height: 2rem;
}

.c6 li button:hover,
.c6 li a:hover,
.c6 li button:focus,
.c6 li a:focus {
  background: #d5d5d5;
}

.c0 {
  overflow: visible;
  position: relative;
}

<div
  className="c0 c1"
>
  <div
    className="c2 c3"
  >
    <button
      className="c4 c5"
      tabIndex={-1}
    >
      show more
    </button>
    <ol
      className="c6"
    >
      <li>
        <button
          onClick={[Function]}
        >
          Delete
        </button>
      </li>
      <li>
        <a
          href="/wooo"
          onClick={[Function]}
        >
          Edit
        </a>
      </li>
    </ol>
  </div>
  <button
    className="c4 c5"
    tabIndex={-1}
  >
    show more
  </button>
</div>
`;

exports[`Dropdown matches snapshot for tab hidden (closed) 1`] = `
.c3 {
  cursor: pointer;
}

.c1 > *:not(.c2) {
  -webkit-animation: 100ms bcCCNc ease-out;
  animation: 100ms bcCCNc ease-out;
  position: absolute;
  box-shadow: 0 0.5rem 0.5rem 0 rgba(0,0,0,0.1);
  border: 1px solid #d5d5d5;
  top: calc(100% + 0.4rem);
  left: 0;
}

.c4 .c5 + .c2 {
  height: 0;
  width: 0;
  overflow: hidden;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c4 .c5.focus-within + .c2 {
  height: unset;
  width: unset;
  -webkit-clip: unset;
  clip: unset;
  overflow: visible;
}

.c4 .c5:focus-within + .c2 {
  height: unset;
  width: unset;
  -webkit-clip: unset;
  clip: unset;
  overflow: visible;
}

.c4 .c5 > .c2 {
  height: unset;
  width: unset;
  -webkit-clip: unset;
  clip: unset;
  overflow: visible;
}

.c4 .c5.focus-within > .c2 {
  height: 0;
  width: 0;
  overflow: hidden;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c4 .c5:focus-within > .c2 {
  height: 0;
  width: 0;
  overflow: hidden;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c4 .c5 > *:not(.c2) {
  -webkit-animation: 100ms bcCCNc ease-out;
  animation: 100ms bcCCNc ease-out;
  position: absolute;
  box-shadow: 0 0.5rem 0.5rem 0 rgba(0,0,0,0.1);
  border: 1px solid #d5d5d5;
  top: calc(100% + 0.4rem);
  left: 0;
}

.c4 .c5 > *:not(.c2) {
  height: 0;
  width: 0;
  overflow: hidden;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c4 .c5.focus-within > *:not(.c2) {
  height: unset;
  width: unset;
  -webkit-clip: unset;
  clip: unset;
  overflow: visible;
}

.c4 .c5:focus-within > *:not(.c2) {
  height: unset;
  width: unset;
  -webkit-clip: unset;
  clip: unset;
  overflow: visible;
}

.c0 {
  overflow: visible;
  position: relative;
}

<div
  className="c0 c1"
>
  <button
    className="c2 c3"
    onClick={[Function]}
  >
    show more
  </button>
</div>
`;

exports[`Dropdown matches snapshot for tab hidden (open) 1`] = `
.c3 {
  cursor: pointer;
}

.c1 > *:not(.c2) {
  -webkit-animation: 100ms bcCCNc ease-out;
  animation: 100ms bcCCNc ease-out;
  position: absolute;
  box-shadow: 0 0.5rem 0.5rem 0 rgba(0,0,0,0.1);
  border: 1px solid #d5d5d5;
  top: calc(100% + 0.4rem);
  left: 0;
}

.c5 .c6 + .c2 {
  height: 0;
  width: 0;
  overflow: hidden;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c5 .c6.focus-within + .c2 {
  height: unset;
  width: unset;
  -webkit-clip: unset;
  clip: unset;
  overflow: visible;
}

.c5 .c6:focus-within + .c2 {
  height: unset;
  width: unset;
  -webkit-clip: unset;
  clip: unset;
  overflow: visible;
}

.c5 .c6 > .c2 {
  height: unset;
  width: unset;
  -webkit-clip: unset;
  clip: unset;
  overflow: visible;
}

.c5 .c6.focus-within > .c2 {
  height: 0;
  width: 0;
  overflow: hidden;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c5 .c6:focus-within > .c2 {
  height: 0;
  width: 0;
  overflow: hidden;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c5 .c6 > *:not(.c2) {
  -webkit-animation: 100ms bcCCNc ease-out;
  animation: 100ms bcCCNc ease-out;
  position: absolute;
  box-shadow: 0 0.5rem 0.5rem 0 rgba(0,0,0,0.1);
  border: 1px solid #d5d5d5;
  top: calc(100% + 0.4rem);
  left: 0;
}

.c5 .c6 > *:not(.c2) {
  height: 0;
  width: 0;
  overflow: hidden;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c5 .c6.focus-within > *:not(.c2) {
  height: unset;
  width: unset;
  -webkit-clip: unset;
  clip: unset;
  overflow: visible;
}

.c5 .c6:focus-within > *:not(.c2) {
  height: unset;
  width: unset;
  -webkit-clip: unset;
  clip: unset;
  overflow: visible;
}

.c4 {
  margin: 0;
  padding: 0.6rem 0;
  background: #f5f5f5;
}

.c4 li button,
.c4 li a {
  text-align: left;
  cursor: pointer;
  outline: none;
  border: none;
  padding: 0 0.8rem;
  margin: 0;
  height: 3.2rem;
  background: none;
  min-width: 7.2rem;
  font-family: Neue Helvetica W01;
  color: #424242;
  font-size: 1.4rem;
  line-height: 2rem;
}

.c4 li button:hover,
.c4 li a:hover,
.c4 li button:focus,
.c4 li a:focus {
  background: #d5d5d5;
}

.c0 {
  overflow: visible;
  position: relative;
}

<div
  className="c0 c1"
>
  <button
    className="c2 c3"
    onClick={[Function]}
  >
    show more
  </button>
  <ol
    className="c4"
  >
    <li>
      <button
        onClick={[Function]}
      >
        Delete
      </button>
    </li>
    <li>
      <a
        href="/wooo"
        onClick={[Function]}
      >
        Edit
      </a>
    </li>
  </ol>
</div>
`;
